src/main/java/seedu/address/model/TagList.java

```diff
@@ -0,0 +1,107 @@
+package seedu.address.model;
+
+import static java.util.Objects.requireNonNull;
+
+import java.util.HashMap;
+import java.util.Set;
+import java.util.TreeSet;
+
+import seedu.address.model.tag.Tag;
+import seedu.address.model.task.Task;
+
+/** A list of unique tags found in the task list */
+public class TagList {
+
+    /**
+     * Hashmap to store tags, where the key is the Tag and the value is an Integer that counts
+     * the number of tasks in the task list which contain the Tag.
+     */
+    private static HashMap<Tag, Integer> tags;
+
+    /**
+     * Creates a tag list, initialised with tags from the existing tasks in the task list.
+     *
+     * @param taskList Task list to retrieve tags from.
+     */
+    public TagList(TaskList taskList) {
+        tags = new HashMap<>();
+        initialiseTagList(taskList);
+    }
+
+    /**
+     * Initialises the tag list with tags found in the given task list.
+     *
+     * @param taskList TaskList to retrieve tags from.
+     */
+    public void initialiseTagList(TaskList taskList) {
+        for (Task task : taskList.getTaskList()) {
+            Set<Tag> tagsInTask = task.getTags();
+            tagsInTask.forEach(this::addTag);
+        }
+    }
+
+    /**
+     * Adds a tag to the tag list.
+     */
+    public void addTag(Tag tag) {
```

Comment:
just a small note, will it be better if this is set to private access instead so that addTag(Tag) and addTag(Task) won't be mixed up? Seems like only addTag(Task) is called outside this class if im not wrong!!